tools/fix_mailmap.py

```diff
@@ -0,0 +1,573 @@
+
+
+def get_raw_authors() -> dict[str, set[tuple]]:
+    """Get the list of authors WITHOUT .mailmap remapping.
+
+    Returns a dict {lowercase_email: set((name, email))}, which will be used to check if a
+    mailmap line is necessary (matches some commit, and actually makes a change)
+    """
+    cmd = [
+        "git",
+        "log",
+        "--pretty=tformat:%an <%ae>%n%(trailers:key=co-authored-by,valueonly=true,unfold=true)",
+    ]
+    res = subprocess.run(cmd, capture_output=True, check=True, text=True)
+
+    logline = re.compile(r"([^<]+)\s+<(.*)>$")
+
+    emails_map: dict[str, set[tuple]] = {}
+    for line in res.stdout.splitlines():
+        match = logline.fullmatch(line.strip())
+        if match:
+            name, email = match.group(1, 2)
+            emails_map.setdefault(email.lower(), set()).add((name, email))
+    return emails_map
+
+
+def get_shortlog_authors() -> tuple[list[tuple], list[str]]:
+    """Get the list of authors from git-shortlog (which uses .mailmap).
+
+    Returns (matches: (num_commits, name, email), non_matches)
+    """
+    cmd = ["git", "shortlog", "-se", "--group=author", "--group=trailer:co-authored-by"]
+    res = subprocess.run(cmd, capture_output=True, check=True, text=True)
+
+    matches = []
+    nonmatches = []
+    shortlogline = re.compile(r"^(\d+)\s+([^<]+)\s+<(.*)>$")
+    for line in res.stdout.splitlines():
+        match = shortlogline.fullmatch(line.strip())
+        if match:
+            matches.append((int(match.group(1)), *match.group(2, 3)))
+        else:
+            # These are usually faulty co-author lines missing <> around the email address eg
+            # Co-Authored-By: firstname lastname email@example.com
+            nonmatches.append(line)
+    return matches, nonmatches
+
+
+def split_mailmap_header_footer(mailmap_file: Path) -> tuple:
+    """Return the mailmap split into header comment lines and body lines."""
+    mailmap = mailmap_file.read_text(encoding="UTF8").splitlines()
+    for i, line in enumerate(mailmap):
+        if not line.startswith("#"):
+            header_end = i
+            break
+    header_lines = mailmap[:header_end]
+    mailmap_lines = mailmap[header_end:]
+    return header_lines, mailmap_lines
+
+
+def parse_mailmap(mailmap_lines: list[str], raw_authors: dict):
+    """Parse the mailmap body, making several checks and fixes.
+
+    1. Warn on non-standard lines
+    2. Remove name-matching
+    3. Drop exact duplicate lines
+    4. Warn for partial duplicate lines
+    5. Drop lines that don't match a commit
+    6. Drop lines that match a commit but cause no changes
+    7. Alphabetise the body
+    8. Warn about multiple canonical names associated with same email (we assume project-unique names)
+    9. Warn about multiple canonical emails associated with same name (we assume project-unique emails)
+    """
+    line_re = re.compile(r"^([^<]+)\s+<([^>]*)>(?:(\s+[^<]+)?\s+<([^>]*)>)?$")
```

Comment:
Given the complexity of this, if the intent is to be a little more maintainable, it may not hurt to use named match groups (which would also mean the `(?:)` isn't needed for logical ordering) like
```python
re.compile(r"""
    (?P<canonical_name>[^<]+)\s+
    <(?P<canonical_email>[^>]*)>
    (\s+(?P<bad_name>[^<]+)?)?
    (\s+<(?P<bad_email>[^>]*)>)?
""", flags=re.X | re.U)
```

(you can also use `re.fullmatch` instead of wrapping the regex in `^` and `$` if you prefer)